### PR TITLE
fix lr_scheduler for continuing training

### PR DESCRIPTION
**Pull request by taoari**
Fix lr_scheduler and update `begin_num_update` for continuing training with `--load-epoch`.

### CODE REVIEW

**Comment by mli** [comment]
@tornadomeet can you please review this pr? it may be related to resnet's retrain

**Comment by tornadomeet** [comment]
ok, i,ll check it today when i go to comp.

**Comment by tornadomeet** [comment]
@mli ```lr_scheduler``` is indeed a problem when retrian, i have used a simple function https://github.com/tornadomeet/ResNet/blob/master/train_resnet.py#L9 when training resnet, so it is ok there.
but i'm not sure whether this pr will indeed fixed it and whether it will influence other part or not.  

 @taoari would you do a toy test, suppose we use ```mx.lr_scheduler.MultiFactorScheduler(step=[1epoch, 3epoch], factor=0.1)```, and start training at epoch 1,2,3,4 respectively, then print out the log, we want to make sure that the logging.info log when changing `lr` is really we want . (we can use a very shallow network in mnist, so this experiment will not cost). thanks!

**Comment by taoari** [comment]
@tornadomeet It is tested. For example, with the parameters:

```
data_dir: ../data/mnist/
dataset: mnist
num_classes: 10
num_examples: 60000
batch_size: 128

lr: 0.01
lr_factor: 0.9
lr_factor_epoch: 1,2,6,7
momentum: 0.8
wd: 0.00005

display: 50
eval_epoch: 2
checkpoint_epoch: 5
num_epochs: 10
model_prefix: mnist_mlp

kv_store: local
network: mlp
initializer: msra
```

We load from epoch 5, and at iteration 2340 (epoch 5), the learning rate is recovered to 0.0081, and for epoch 6 and 7, the lr_farctor is multiplied. 

The log file is attached.

```
2016-09-10 16:10:38,789 Node[0] Running on machine: 4YRK2Y1
2016-09-10 16:10:38,806 Node[0] loading from mnist_mlp-0-0005.params
2016-09-10 16:10:41,269 Node[0] lr_mult: []
2016-09-10 16:10:41,269 Node[0] wd_mult: [('fc1_bias', 0.0), ('fc2_bias', 0.0), ('fc3_bias', 0.0)]
2016-09-10 16:10:41,270 Node[0] Start training with [cpu(0)]
2016-09-10 16:10:41,273 Node[0] Update[2340]: Change learning rate to 9.00000e-03
2016-09-10 16:10:41,273 Node[0] Update[2340]: Change learning rate to 8.10000e-03
2016-09-10 16:10:41,275 Node[0] Epoch[4] lr = 0.008100
2016-09-10 16:10:41,335 Node[0] Epoch[4] Validation-accuracy=0.952925
2016-09-10 16:10:41,336 Node[0] Epoch[4] Validation-cross-entropy=0.157290
2016-09-10 16:10:41,430 Node[0] Epoch[5] Batch [50]	Speed: 70329.61 samples/sec	Train-accuracy=0.958125
2016-09-10 16:10:41,431 Node[0] Epoch[5] Batch [50]	Speed: 70329.61 samples/sec	Train-cross-entropy=0.153342
2016-09-10 16:10:41,522 Node[0] Epoch[5] Batch [100]	Speed: 71110.99 samples/sec	Train-accuracy=0.957500
2016-09-10 16:10:41,523 Node[0] Epoch[5] Batch [100]	Speed: 71110.99 samples/sec	Train-cross-entropy=0.158887
2016-09-10 16:10:41,612 Node[0] Epoch[5] Batch [150]	Speed: 71910.12 samples/sec	Train-accuracy=0.954688
2016-09-10 16:10:41,614 Node[0] Epoch[5] Batch [150]	Speed: 71910.12 samples/sec	Train-cross-entropy=0.161700
2016-09-10 16:10:41,704 Node[0] Epoch[5] Batch [200]	Speed: 70329.79 samples/sec	Train-accuracy=0.955781
2016-09-10 16:10:41,706 Node[0] Epoch[5] Batch [200]	Speed: 70329.79 samples/sec	Train-cross-entropy=0.145335
2016-09-10 16:10:41,799 Node[0] Epoch[5] Batch [250]	Speed: 69565.21 samples/sec	Train-accuracy=0.956875
2016-09-10 16:10:41,799 Node[0] Epoch[5] Batch [250]	Speed: 69565.21 samples/sec	Train-cross-entropy=0.153545
2016-09-10 16:10:41,885 Node[0] Epoch[5] Batch [300]	Speed: 74418.63 samples/sec	Train-accuracy=0.954688
2016-09-10 16:10:41,887 Node[0] Epoch[5] Batch [300]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.152972
2016-09-10 16:10:41,973 Node[0] Epoch[5] Batch [350]	Speed: 74418.63 samples/sec	Train-accuracy=0.957500
2016-09-10 16:10:41,974 Node[0] Epoch[5] Batch [350]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.147508
2016-09-10 16:10:42,062 Node[0] Epoch[5] Batch [400]	Speed: 74418.63 samples/sec	Train-accuracy=0.960313
2016-09-10 16:10:42,062 Node[0] Epoch[5] Batch [400]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.147308
2016-09-10 16:10:42,148 Node[0] Epoch[5] Batch [450]	Speed: 75294.08 samples/sec	Train-accuracy=0.956719
2016-09-10 16:10:42,150 Node[0] Epoch[5] Batch [450]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.152103
2016-09-10 16:10:42,180 Node[0] Epoch[5] Resetting Data Iterator
2016-09-10 16:10:42,180 Node[0] Epoch[5] Time cost=0.843
2016-09-10 16:10:42,181 Node[0] Epoch[5] lr = 0.008100
2016-09-10 16:10:42,236 Node[0] Epoch[5] Validation-accuracy=0.956530
2016-09-10 16:10:42,236 Node[0] Epoch[5] Validation-cross-entropy=0.144011
2016-09-10 16:10:42,237 Node[0] Update[2809]: Change learning rate to 7.29000e-03
2016-09-10 16:10:42,322 Node[0] Epoch[6] Batch [50]	Speed: 77108.49 samples/sec	Train-accuracy=0.963281
2016-09-10 16:10:42,323 Node[0] Epoch[6] Batch [50]	Speed: 77108.49 samples/sec	Train-cross-entropy=0.138273
2016-09-10 16:10:42,410 Node[0] Epoch[6] Batch [100]	Speed: 75294.08 samples/sec	Train-accuracy=0.961719
2016-09-10 16:10:42,411 Node[0] Epoch[6] Batch [100]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.143792
2016-09-10 16:10:42,499 Node[0] Epoch[6] Batch [150]	Speed: 73563.11 samples/sec	Train-accuracy=0.956719
2016-09-10 16:10:42,500 Node[0] Epoch[6] Batch [150]	Speed: 73563.11 samples/sec	Train-cross-entropy=0.146354
2016-09-10 16:10:42,586 Node[0] Epoch[6] Batch [200]	Speed: 74418.63 samples/sec	Train-accuracy=0.960625
2016-09-10 16:10:42,588 Node[0] Epoch[6] Batch [200]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.129815
2016-09-10 16:10:42,674 Node[0] Epoch[6] Batch [250]	Speed: 74418.63 samples/sec	Train-accuracy=0.960625
2016-09-10 16:10:42,676 Node[0] Epoch[6] Batch [250]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.138354
2016-09-10 16:10:42,763 Node[0] Epoch[6] Batch [300]	Speed: 74418.63 samples/sec	Train-accuracy=0.957656
2016-09-10 16:10:42,763 Node[0] Epoch[6] Batch [300]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.138450
2016-09-10 16:10:42,849 Node[0] Epoch[6] Batch [350]	Speed: 75294.30 samples/sec	Train-accuracy=0.961406
2016-09-10 16:10:42,851 Node[0] Epoch[6] Batch [350]	Speed: 75294.30 samples/sec	Train-cross-entropy=0.133973
2016-09-10 16:10:42,940 Node[0] Epoch[6] Batch [400]	Speed: 72727.22 samples/sec	Train-accuracy=0.965000
2016-09-10 16:10:42,940 Node[0] Epoch[6] Batch [400]	Speed: 72727.22 samples/sec	Train-cross-entropy=0.133789
2016-09-10 16:10:43,026 Node[0] Epoch[6] Batch [450]	Speed: 74418.63 samples/sec	Train-accuracy=0.961250
2016-09-10 16:10:43,028 Node[0] Epoch[6] Batch [450]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.138178
2016-09-10 16:10:43,059 Node[0] Epoch[6] Resetting Data Iterator
2016-09-10 16:10:43,061 Node[0] Epoch[6] Time cost=0.824
2016-09-10 16:10:43,062 Node[0] Epoch[6] lr = 0.007290
2016-09-10 16:10:43,063 Node[0] Update[3277]: Change learning rate to 6.56100e-03
2016-09-10 16:10:43,151 Node[0] Epoch[7] Batch [50]	Speed: 75294.08 samples/sec	Train-accuracy=0.965000
2016-09-10 16:10:43,151 Node[0] Epoch[7] Batch [50]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.126451
2016-09-10 16:10:43,242 Node[0] Epoch[7] Batch [100]	Speed: 72727.42 samples/sec	Train-accuracy=0.966406
2016-09-10 16:10:43,242 Node[0] Epoch[7] Batch [100]	Speed: 72727.42 samples/sec	Train-cross-entropy=0.132558
2016-09-10 16:10:43,336 Node[0] Epoch[7] Batch [150]	Speed: 68817.25 samples/sec	Train-accuracy=0.962187
2016-09-10 16:10:43,338 Node[0] Epoch[7] Batch [150]	Speed: 68817.25 samples/sec	Train-cross-entropy=0.134527
2016-09-10 16:10:43,424 Node[0] Epoch[7] Batch [200]	Speed: 73563.31 samples/sec	Train-accuracy=0.965000
2016-09-10 16:10:43,426 Node[0] Epoch[7] Batch [200]	Speed: 73563.31 samples/sec	Train-cross-entropy=0.117938
2016-09-10 16:10:43,512 Node[0] Epoch[7] Batch [250]	Speed: 75294.08 samples/sec	Train-accuracy=0.962656
2016-09-10 16:10:43,513 Node[0] Epoch[7] Batch [250]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.126988
2016-09-10 16:10:43,598 Node[0] Epoch[7] Batch [300]	Speed: 75294.08 samples/sec	Train-accuracy=0.961719
2016-09-10 16:10:43,599 Node[0] Epoch[7] Batch [300]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.127419
2016-09-10 16:10:43,687 Node[0] Epoch[7] Batch [350]	Speed: 73563.11 samples/sec	Train-accuracy=0.964375
2016-09-10 16:10:43,687 Node[0] Epoch[7] Batch [350]	Speed: 73563.11 samples/sec	Train-cross-entropy=0.123390
2016-09-10 16:10:43,801 Node[0] Epoch[7] Batch [400]	Speed: 56637.09 samples/sec	Train-accuracy=0.968437
2016-09-10 16:10:43,802 Node[0] Epoch[7] Batch [400]	Speed: 56637.09 samples/sec	Train-cross-entropy=0.123540
2016-09-10 16:10:43,894 Node[0] Epoch[7] Batch [450]	Speed: 70329.61 samples/sec	Train-accuracy=0.964688
2016-09-10 16:10:43,894 Node[0] Epoch[7] Batch [450]	Speed: 70329.61 samples/sec	Train-cross-entropy=0.127489
2016-09-10 16:10:43,926 Node[0] Epoch[7] Resetting Data Iterator
2016-09-10 16:10:43,926 Node[0] Epoch[7] Time cost=0.863
2016-09-10 16:10:43,927 Node[0] Epoch[7] lr = 0.006561
2016-09-10 16:10:43,980 Node[0] Epoch[7] Validation-accuracy=0.961739
2016-09-10 16:10:43,980 Node[0] Epoch[7] Validation-cross-entropy=0.126572
2016-09-10 16:10:44,068 Node[0] Epoch[8] Batch [50]	Speed: 76190.38 samples/sec	Train-accuracy=0.967656
2016-09-10 16:10:44,069 Node[0] Epoch[8] Batch [50]	Speed: 76190.38 samples/sec	Train-cross-entropy=0.117629
2016-09-10 16:10:44,154 Node[0] Epoch[8] Batch [100]	Speed: 75294.08 samples/sec	Train-accuracy=0.968437
2016-09-10 16:10:44,154 Node[0] Epoch[8] Batch [100]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.124062
2016-09-10 16:10:44,242 Node[0] Epoch[8] Batch [150]	Speed: 74418.43 samples/sec	Train-accuracy=0.964375
2016-09-10 16:10:44,243 Node[0] Epoch[8] Batch [150]	Speed: 74418.43 samples/sec	Train-cross-entropy=0.125556
2016-09-10 16:10:44,332 Node[0] Epoch[8] Batch [200]	Speed: 72727.22 samples/sec	Train-accuracy=0.967344
2016-09-10 16:10:44,332 Node[0] Epoch[8] Batch [200]	Speed: 72727.22 samples/sec	Train-cross-entropy=0.109068
2016-09-10 16:10:44,421 Node[0] Epoch[8] Batch [250]	Speed: 71910.12 samples/sec	Train-accuracy=0.965000
2016-09-10 16:10:44,421 Node[0] Epoch[8] Batch [250]	Speed: 71910.12 samples/sec	Train-cross-entropy=0.118240
2016-09-10 16:10:44,507 Node[0] Epoch[8] Batch [300]	Speed: 75294.30 samples/sec	Train-accuracy=0.963437
2016-09-10 16:10:44,509 Node[0] Epoch[8] Batch [300]	Speed: 75294.30 samples/sec	Train-cross-entropy=0.118894
2016-09-10 16:10:44,595 Node[0] Epoch[8] Batch [350]	Speed: 75294.08 samples/sec	Train-accuracy=0.966875
2016-09-10 16:10:44,595 Node[0] Epoch[8] Batch [350]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.115051
2016-09-10 16:10:44,684 Node[0] Epoch[8] Batch [400]	Speed: 73563.11 samples/sec	Train-accuracy=0.970938
2016-09-10 16:10:44,684 Node[0] Epoch[8] Batch [400]	Speed: 73563.11 samples/sec	Train-cross-entropy=0.115458
2016-09-10 16:10:44,770 Node[0] Epoch[8] Batch [450]	Speed: 74418.63 samples/sec	Train-accuracy=0.966719
2016-09-10 16:10:44,772 Node[0] Epoch[8] Batch [450]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.119168
2016-09-10 16:10:44,803 Node[0] Epoch[8] Resetting Data Iterator
2016-09-10 16:10:44,803 Node[0] Epoch[8] Time cost=0.822
2016-09-10 16:10:44,805 Node[0] Epoch[8] lr = 0.006561
2016-09-10 16:10:44,892 Node[0] Epoch[9] Batch [50]	Speed: 76190.59 samples/sec	Train-accuracy=0.969844
2016-09-10 16:10:44,894 Node[0] Epoch[9] Batch [50]	Speed: 76190.59 samples/sec	Train-cross-entropy=0.110120
2016-09-10 16:10:44,980 Node[0] Epoch[9] Batch [100]	Speed: 75294.08 samples/sec	Train-accuracy=0.970469
2016-09-10 16:10:44,980 Node[0] Epoch[9] Batch [100]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.116483
2016-09-10 16:10:45,068 Node[0] Epoch[9] Batch [150]	Speed: 74418.63 samples/sec	Train-accuracy=0.966406
2016-09-10 16:10:45,069 Node[0] Epoch[9] Batch [150]	Speed: 74418.63 samples/sec	Train-cross-entropy=0.117605
2016-09-10 16:10:45,154 Node[0] Epoch[9] Batch [200]	Speed: 75294.08 samples/sec	Train-accuracy=0.970000
2016-09-10 16:10:45,155 Node[0] Epoch[9] Batch [200]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.101319
2016-09-10 16:10:45,240 Node[0] Epoch[9] Batch [250]	Speed: 76190.59 samples/sec	Train-accuracy=0.968281
2016-09-10 16:10:45,242 Node[0] Epoch[9] Batch [250]	Speed: 76190.59 samples/sec	Train-cross-entropy=0.110633
2016-09-10 16:10:45,329 Node[0] Epoch[9] Batch [300]	Speed: 73563.31 samples/sec	Train-accuracy=0.965781
2016-09-10 16:10:45,331 Node[0] Epoch[9] Batch [300]	Speed: 73563.31 samples/sec	Train-cross-entropy=0.111314
2016-09-10 16:10:45,415 Node[0] Epoch[9] Batch [350]	Speed: 76190.59 samples/sec	Train-accuracy=0.969844
2016-09-10 16:10:45,417 Node[0] Epoch[9] Batch [350]	Speed: 76190.59 samples/sec	Train-cross-entropy=0.107656
2016-09-10 16:10:45,502 Node[0] Epoch[9] Batch [400]	Speed: 75294.08 samples/sec	Train-accuracy=0.973125
2016-09-10 16:10:45,503 Node[0] Epoch[9] Batch [400]	Speed: 75294.08 samples/sec	Train-cross-entropy=0.108214
2016-09-10 16:10:45,589 Node[0] Epoch[9] Batch [450]	Speed: 73563.31 samples/sec	Train-accuracy=0.969063
2016-09-10 16:10:45,591 Node[0] Epoch[9] Batch [450]	Speed: 73563.31 samples/sec	Train-cross-entropy=0.111774
2016-09-10 16:10:45,621 Node[0] Epoch[9] Resetting Data Iterator
2016-09-10 16:10:45,622 Node[0] Epoch[9] Time cost=0.816
2016-09-10 16:10:45,624 Node[0] Epoch[9] lr = 0.006561
2016-09-10 16:10:45,714 Node[0] Saved checkpoint to "mnist_mlp-0-0010.params"
2016-09-10 16:10:45,811 Node[0] Epoch[9] Validation-accuracy=0.965345
2016-09-10 16:10:45,812 Node[0] Epoch[9] Validation-cross-entropy=0.114606
2016-09-10 16:10:45,812 Node[0] Optimization done.
```

**Comment by tornadomeet** [comment]
@mli LGTM now.

@taoari thanks!  could you using `git rebase` to combine all the commit to one, so that it is a meaningfule commit. Please refer http://mxnet.readthedocs.io/en/latest/how_to/contribute.html#how-to-combine-multiple-commits-into-one
